text/0000-mergeable-rustdoc-cross-crate-info.md

```diff
@@ -0,0 +1,441 @@
+
+* `--merge=auto` (`read_rendered_cci && write_rendered_cci`) is the default, and reflects the current behavior of rustdoc. 
+* `--merge=none` (`!read_rendered_cci && !write_rendered_cci`) means that rustdoc will ignore the cross-crate files in the doc root. Only generate item docs. 
+* `--merge=write-only` (`!read_rendered_cci && write_rendered_cci`) outputs crate info based only on the current crate and `--include-info-json`'ed crates.
+* A (`read_rendered_cci && !write_rendered_cci`) mode would be useless, since the data that is read would be ignored and not written.
+
+## Merge step
+
+This proposal is capable of addressing two primary use cases. It allows developers to enable CCI in these scenarios:
+* Documenting a crate and its transitive dependencies in parallel in build systems that require build actions to be independent
+* Producing a documentation index of a large number of crates, in such a way that if one crate is updated, only the updated crates and an index have to be redocumented. This scenario is demonstrated in the Guide-level explanation.
+
+CCI is not automatically enabled in either situation. A combination of the `--include-info-json`, `--merge`, and `--write-info-json` flags are needed to produce this behavior. This RFC provides a minimal set of tools that allow developers of build systems, like Bazel and Buck2, to create rules for these scenarios.
+
+Discussion of whether additional features should be included to facilitate this merge step can be found in Unresolved questions (Index crate).
+
+## Compatibility
+
+This RFC does not alter previous compatibility guarantees made about the output of rustdoc. In particular it does not stabilize the presence of the rendered cross-crate information files, their content, or the HTML generated by rustdoc.
+
+In the same way that the [rustdoc HTML output is unstable](https://rust-lang.github.io/rfcs/2963-rustdoc-json.html#:~:text=The%20HTML%20output%20of%20rustdoc,into%20a%20different%20format%20impractical), the contents of the `crate-info.json` will be considered unstable. Between versions of rustdoc, breaking changes to the content of `crate-info.json` should be expected. Only the presence of a `crate-info.json` file is promised, under `--write-info-json`. Merging cross-crate information generated by disparate versions of rustdoc is not supported.
+
+The implementation of the RFC itself is designed to produce only minimal changes to cross-crate info files and the HTML output of rustdoc. Exhaustively, the implementation is allowed to 
+* Change the sorting order of trait implementations, type implementations, and other cross-crate info in the HTML output of rustdoc
+* Add a comment on the last line of generated HTML pages, to store metadata relevant to appending items to them
```

Comment:
I'm not sure to see the need for this requirement. Do you have an example in mind?

Reply:
I don't want to make any changes to the existing tools that use rustdoc across multiple crates by documenting every crate in the same output directory. If we stop supporting this mode, there are a lot of tools that we would break, including the rustdoc test suite, Cargo, and numerous other proprietary tools. That's why the RFC and my implementation has tighter requirements on what it's going to do, and that's why it has a default `--mode=read-write`.

I expect the sorting order of cross-crate information to change because the existing sorting order depends on specific implementation details, like whether we're sorting by the serialized or native representations of the CCI.

I am allowing adding a comment to the last line of the rendered files, because this is how my implementation implements `read_rendered_cci`. It finds out where to insert the next part into an existing file by looking at a comment that is written to the last line.

Reply:
Hum. There is already [`--html-after-content`](https://doc.rust-lang.org/rustdoc/command-line-arguments.html?#--html-after-content-include-more-html-after-the-content). Wouldn't it cover this use case?